cpp/src/arrow/compare.cc

```diff
@@ -1336,4 +1336,52 @@ bool TypeEquals(const DataType& left, const DataType& right, bool check_metadata
   }
 }
 
+#define ARROW_CMP_TYPE_CASE(id, T, Op)                           \
+  case Type::id:                                                 \
+    return internal::NumericScalarCompare<NumericScalar<T>, Op>( \
+        static_cast<const NumericScalar<T>&>(left),              \
+        static_cast<const NumericScalar<T>&>(right), options);
```

Comment:
Should use `checked_cast` so that debug builds actually check the type downcast runtime (`checked_cast` is turned into `dynamic_cast` in debug mode and `static_cast` in release mode).